File: apps/ios/Wordocious/Sources/PresenceService.swift
import Foundation
import SocketIO

/// Always-on lightweight presence socket — the native counterpart to the web
/// `SitePresenceProvider`. Holds one Socket.IO connection while the app is
/// foregrounded so the signed-in user is counted in the server's `/presence`
/// total (what the home LIVE banner renders). Tagged with the SAME
/// `u:<userId>` presenceId the VS match socket uses (VSMatchViewModel), so the
/// server dedupes a person to 1 even while they're in a match. Kept separate
/// from VSMatchService so it never touches matchmaking state.
@MainActor
final class PresenceService {
    static let shared = PresenceService()
    private init() {}

    private var manager: SocketManager?
    private var socket: SocketIOClient?

    /// Matches web `getPresenceId()` + VSMatchViewModel: `u:<supabase-user-id>`.
    private var presenceId: String? {
        AuthService.shared.profile.map { "u:\($0.id)" }
    }

    /// Idempotent: no-ops if already connected or the user isn't loaded yet.
    func start() {
        guard socket == nil, VSConfig.isConfigured,
              let url = VSConfig.serverURL, let pid = presenceId else { return }
        let manager = SocketManager(socketURL: url, config: [
            .log(false), .compress, .reconnects(true), .reconnectWait(2), .reconnectWaitMax(10),
        ])
        self.manager = manager
        let socket = manager.defaultSocket
        self.socket = socket
        // No event handlers needed — the server counts the connection itself.
        socket.connect(withPayload: ["presenceId": pid])
    }

    func stop() {
        socket?.disconnect()
        manager?.disconnect()
        socket = nil
        manager = nil
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/PresenceService.kt
package com.wordocious.app.data

import io.socket.client.IO
import io.socket.client.Socket

/**
 * Always-on lightweight presence socket — the Android counterpart to the web
 * SitePresenceProvider / iOS PresenceService. Holds one socket.io connection
 * while signed in so the user is counted in the server's /presence total (the
 * home LIVE banner). Tagged with the SAME `u:<userId>` presenceId the VS match
 * socket uses, so the server dedupes a person to 1 even while in a match.
 */
object PresenceService {
    private var socket: Socket? = null

    private val presenceId: String? get() = AuthService.userId?.let { "u:$it" }

    /** Idempotent: no-ops if already connected or the user isn't loaded yet. */
    fun start() {
        if (socket != null || !VSConfig.isConfigured) return
        val pid = presenceId ?: return
        val opts = IO.Options().apply {
            reconnection = true
            transports = arrayOf("websocket")
            auth = mapOf("presenceId" to pid)
        }
        val s = runCatching { IO.socket(VSConfig.SERVER_URL, opts) }.getOrNull() ?: return
        socket = s
        s.connect()   // no handlers needed — the server counts the connection itself
    }

    fun stop() {
        socket?.disconnect()
        socket = null
    }
}
